Introduce PTES and TES constraints

TomKae00 opened the pull request:
## Changes proposed in this Pull Request

This pull request introduces the following adjustments to the ``` prepare_sector_network ``` and ``` solve_network ``` scripts, replacing PR #1444 (which did not work with the HiGHS solver):

- Added PTES.
- Introduced a constraint to enforce the energy-to-power ratio (etpr).
- Implemented a constraint to ensure that the charger and discharger of the thermal energy stores are sized equally.

The following config data has been used:

```yaml
scenario:
  clusters:
    - 6

planning_horizons:
  - 2030

foresight: overnight
countries: ['DE', 'DK']

clustering:
  temporal:
    resolution_sector: 12h
```

## Constraints

The etpr ratio constraint and the constraint enforcing equal sizing of the charger and discharger are working as intended, as shown in the figure below. In particular, the size of the store divided by 150 equals the size of both the charger and the discharger.

<img src="https://github.com/user-attachments/assets/38c6b065-335c-49a0-a517-476f45d1ccf1" alt="charger_discharger_profile_comparison" width="700"/>

The annual charging and discharging process for an exemplary PTES store is illustrated in the figure below:

<img src="https://github.com/user-attachments/assets/e648300e-5162-4ad7-be75-e5c15b4d78b6" alt="charger_discharger_size_comparison" width="700" />


## System Costs

System costs have increased slightly with the newly introduced constraints. This is logical given the larger capacities required when the charger and discharger can no longer be chosen independently.

<img src="https://github.com/user-attachments/assets/1fbd0931-d1f5-4c77-ad92-15a8f20d9b3a" alt="total_cost_comparison" width="700" />




## Checklist

- [ x] I tested my contribution locally and it works as intended.
- [ x] Code and workflow changes are sufficiently documented.
- [ x] Changed dependencies are added to `envs/environment.yaml`.
- [ x] Changes in configuration options are added in `config/config.default.yaml`.
- [ x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [ x] Sources of newly added data are documented in `doc/data_sources.rst`.
- [ x] A release note `doc/release_notes.rst` is added.

amos-schledorn reviewed (changes requested):
Looks great! Just a few small comments. When the CI runs through, I'll test this locally.

TomKae00 commented:
@amos-schledorn thanks for the review! I've implemented your suggestions. Moreover, i added error handling to ensure that the TES constraints are only applied when intended.

TomKae00 commented:
Hi @fneum and @amos-schledorn,

Currently, the CI is failing when running:

``snakemake make_summary_perfect --configfile config/test/config.perfect.yaml``

**Reason for CI Failure:**
In the master, the water tank charger and discharger are included in costs.csv with an efficiency of 0.9. Because of this, they can be used as heat vents, allowing energy to be released from the system.

With this PR, the efficiency of both the central/decentral chargers and dischargers has been set to 1.0 according to the DEA catalogue. As a result, they can no longer act as heat vents, which prevents energy from being released. Additionally, chargers are now directly tied to the store size via the energy-to-power ratio, meaning they cannot be used independently.

**Possible Solution:**
Since energy can no longer be released through the chargers/dischargers, the system costs are now higher. A potential solution could be to allow heat vents to be built for all heat systems, as it is done for urban central heating systems in ``prepare_sector_network.py`` line 2437.

Is there a specific reason why heat vents are currently only allowed for urban central heating? Or are there any concerns against extending this option to all systems?

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -794,6 +795,104 @@ def add_operational_reserve_margin(n, sns, config):
+        indices_stores_e_nom_extendable,
+        energy_to_power_ratio_values,
+    ):
+        charger_var = n.model["Link-p_nom"].loc[charger]
```
```suggestion
        charger_var = n.model["Link-p_nom"].loc[charger]
        if not tes.split(" ")[:4] == charger.split(" ")[:4]:
            # e.g. "DE0 0 urban central water tanks charger" -> ["DE0", "0", "urban", "central", "water", "tanks"]
            raise RuntimeError(
                f"Charger {charger} and TES {tes} do not match. "
                "Ensure that the charger and TES are in the same location and refer to the same technology."
            )
```

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -794,6 +795,104 @@ def add_operational_reserve_margin(n, sns, config):
+    ------
+    ValueError
+        If no valid TES discharger or charger links are found.
+    """
```
```suggestion
        RuntimeError
            If the charger and discharger indices do not align.
    """
```

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -794,6 +795,104 @@ def add_operational_reserve_margin(n, sns, config):
+        raise ValueError(
+            "No valid extendable TES discharger or charger links found for TES charger ratio constraints."
+        )
+
```
```suggestion
    
    for charger, discharger in zip(
        indices_charger_p_nom_extendable, indices_discharger_p_nom_extendable
    ):
        if not charger.split(" ")[:5] == discharger.split(" ")[:5]:
            # e.g. "DE0 0 urban central water tanks charger" -> ["DE0", "0", "urban", "central", "water", "tanks"]
            raise RuntimeError(
                f"Charger {charger} and discharger {discharger} do not match. "
                "Ensure that the charger and discharger are in the same location and refer to the same technology."
            )
    
```

amos-schledorn reviewed (changes requested):
@TomKae00 looks great! I've just suggested a few changes to address the Copilot comment. Let's table this until the heat-venting is fixed.

amos-schledorn commented on scripts/prepare_sector_network.py:
```diff
@@ -2348,6 +2348,7 @@ def add_heat(
     spatial: object,
     options: dict,
     investment_year: int,
+    marginal_cost_water_tank_charger: float,
```
```suggestion
```

amos-schledorn commented on scripts/prepare_sector_network.py:
```diff
@@ -2682,6 +2701,66 @@ def add_heat(
+                    e_nom_extendable=True,
+                    carrier=f"{heat_system} water pits",
+                    standing_loss=1 - np.exp(-1 / 24 / tes_time_constant_days),
+                    capital_cost=costs.at["central water pit storage", "fixed"],
```
```suggestion
                    capital_cost=costs.at["central water pit storage", "capital_cost"],
```

amos-schledorn commented on scripts/prepare_sector_network.py:
```diff
@@ -5258,6 +5337,9 @@ def add_enhanced_geothermal(
             spatial=spatial,
             options=options,
             investment_year=investment_year,
+            marginal_cost_water_tank_charger=snakemake.params.sector[
```
```suggestion
```

amos-schledorn commented on scripts/prepare_sector_network.py:
```diff
@@ -5258,6 +5337,9 @@ def add_enhanced_geothermal(
             options=options,
             investment_year=investment_year,
+            marginal_cost_water_tank_charger=snakemake.params.sector[
+                "marginal_cost_water_tank_charger"
```
```suggestion
```

amos-schledorn commented on scripts/prepare_sector_network.py:
```diff
@@ -5258,6 +5337,9 @@ def add_enhanced_geothermal(
             investment_year=investment_year,
+            marginal_cost_water_tank_charger=snakemake.params.sector[
+                "marginal_cost_water_tank_charger"
+            ],
```
```suggestion
```

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -1189,7 +1335,7 @@ def solve_network(
         snakemake = mock_snakemake(
             "solve_sector_network_perfect",
-            configfiles="../config/test/config.perfect.yaml",
+            configfiles="config/test/config.perfect.yaml",
```
```suggestion
```

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -820,6 +821,124 @@ def add_operational_reserve_margin(n, sns, config):
+        energy_to_power_ratio_values,
+    ):
+        charger_var = n.model["Link-p_nom"].loc[charger]
+        if not tes.replace("-", " ").split(" ")[:5] == charger.split(" ")[:5]:
+            # e.g. "DE0 0 urban central water tanks charger" -> ["DE0", "0", "urban", "central", "water"]
```
Valid comment, but possible issues are addressed by throwing an error when respective string matching fails.

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -820,6 +821,124 @@ def add_operational_reserve_margin(n, sns, config):
+            "No valid extendable TES discharger or charger links found for TES charger ratio constraints."
+        )
+
+    for charger, discharger in zip(
+        indices_charger_p_nom_extendable, indices_discharger_p_nom_extendable
+    ):
```
Valid comment, but possible issues are addressed by throwing an error when respective string matching fails.

cpschau commented on scripts/solve_network.py:
```diff
@@ -1023,6 +1142,33 @@ def extra_functionality(
+            add_TES_energy_to_power_ratio_constraints(n)
+            add_TES_charger_ratio_constraints(n)
+        elif (
+            n.links.index.str.lower()
+            .str.contains("pits charger|tanks charger", case=False, na=False)
```
Why `lower()` and `contains(case=False)` have to be used?

cpschau commented on scripts/solve_network.py:
```diff
@@ -1023,6 +1142,33 @@ def extra_functionality(
 
+    if n.config.get("sector", {}).get("tes", False):
+        if (
+            n.buses.index.str.lower()
+            .str.contains(
+                r"urban central heat|urban decentral heat|rural heat",
+                case=False,
```
Why `lower()` and `contains(case=False)` have to be used?

cpschau reviewed (commented):
Hey @TomKae00,
only had a couple of questions, where I am not sure if they need to be addressed with new changes in the code. Did you discuss the addition of the "energy to power" column with @fneum?

amos-schledorn commented on scripts/solve_network.py:
```diff
@@ -820,6 +821,124 @@ def add_operational_reserve_margin(n, sns, config):
+            )
+
+    eff_discharger = n.links.efficiency[indices_discharger_p_nom_extendable].values
+    lhs = (
```
Hm, I think this is correct based on the equivalent constraint for batteries:
```Add constraint ensuring that charger = discharger, i.e.
    1 * charger_size - efficiency * discharger_size = 0```
